zebedee-cms/src/main/java/com/github/onsdigital/zebedee/keyring/io/CollectionKeyStoreImpl.java

```diff
@@ -0,0 +1,182 @@
+    static final String COLLECTION_KEY_NOT_FOUND_ERR = "collectionKey not found";
+    static final String KEY_DECRYPTION_ERR = "error while decrypting collectionKey file";
+    static final String WRITE_KEY_ERR = "error writing collection key to store";
+    static final String COLLECTION_KEY_ID_NULL_ERR = "collectionKey.ID required but was null or empty";
+    static final String COLLECTION_KEY_KEY_NULL_ERR = "collectionKey.secretKey required but was null";
+    static final String COLLECTION_KEY_ALREADY_EXISTS_ERR = "collectionKey for this collection ID already exists";
+    static final String ALGORITHM = "AES/CBC/PKCS5Padding";
+
+    private Path keyringDir;
+    private Gson gson;
+    private transient SecretKey masterKey;
+    private transient IvParameterSpec masterIv;
+
+    /**
+     * <p>
+     * Create a new CollectionKeyStore instance.
+     * </p>
+     *
+     * <p>
+     * <b>Important:</b> It is strongly recommended that the {@link SecretKey} & {@link IvParameterSpec}
+     * are not modified once they have been used to store collection keys. Changing these inputs will result in this
+     * object no longer being able to decrypt collection key files written with the previous encryption key.
+     * <p>
+     * Reverting the key/init vector values will resolve this issue for keys written with the original encryption
+     * parameters however, the same issue will now be the case for any key files written using the modified encryption
+     * params.
+     * </p>
+     *
+     * <p>
+     * <b>
+     * If a encryption parameters change is necessary it is strongly recommended you migrate all existing keys to
+     * use the updated encryption values.</b>
+     * </p>
+     *
+     * @param keyringDir the {@link Path} the read the {@link CollectionKey} files from.
+     * @param masterKey  the {@link SecretKey} to use when decrypting the {@link CollectionKey} files.
+     * @param masterIv   the {@link IvParameterSpec} to use when initializing the encryption {@link Cipher}.
+     */
+    public CollectionKeyStoreImpl(final Path keyringDir, final SecretKey masterKey, final IvParameterSpec masterIv) {
+        this.keyringDir = keyringDir;
+        this.masterKey = masterKey;
+        this.masterIv = masterIv;
+        this.gson = new GsonBuilder()
+                .registerTypeAdapter(CollectionKey.class, new CollectionKeySerializer())
+                .create();
+    }
+
+    @Override
+    public CollectionKey read(String collectionID) throws KeyringException {
+        if (isEmpty(collectionID)) {
+            throw new KeyringException(INVALID_COLLECTION_ID_ERR);
+        }
+
+        Path keyPath = Paths.get(getKeyPath(collectionID));
+
+        if (Files.notExists(keyPath)) {
+            throw new KeyringException(COLLECTION_KEY_NOT_FOUND_ERR, collectionID);
+        }
+
+        try (
+                RandomAccessFile raf = new RandomAccessFile(getKeyPath(collectionID), "rwd");
```

Comment:
This can be opened as read only?
```suggestion
                RandomAccessFile raf = new RandomAccessFile(getKeyPath(collectionID), "r");
```

Reply:
The access mode is for the lock - as I understand it `rw` mode is an exclusive lock so no one can read the file while this lock has been acquired - which for this is the functionality we need. Whenever we are reading a key file we want to ensure no one else is reading or deleting it.